### ISSUE TITLE
Dropout layer is broken

### ISSUE THREAD

**Issue by randolf-scholz** [tensorflow/tensorflow#25175]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): NO
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.10
- TensorFlow installed from (source or binary): conda default channel
- TensorFlow version (use command below): 1.12 (conda version mkl_py36h69b6ba0_0)
- Python version: 3.6.8

**Describe the current behavior**
The Dropout layer always acts as if it is in testing mode.

**Describe the expected behavior**
The Dropout layer switches between training and testing.

**Code to reproduce the issue**
The tutorial code available from the [hompage](https://www.tensorflow.org/tutorials/):

    import tensorflow as tf
    mnist = tf.keras.datasets.mnist
    
    (x_train, y_train),(x_test, y_test) = mnist.load_data()
    x_train, x_test = x_train / 255.0, x_test / 255.0
    
    model = tf.keras.models.Sequential([
      tf.keras.layers.Flatten(),
      tf.keras.layers.Dense(512, activation=tf.nn.relu),
      tf.keras.layers.Dropout(0.2),
      tf.keras.layers.Dense(10, activation=tf.nn.softmax)
    ])
    model.compile(optimizer='adam',
                  loss='sparse_categorical_crossentropy',
                  metrics=['accuracy'])
    
    model.fit(x_train, y_train, epochs=5)
    model.evaluate(x_test, y_test)

Setting the droprate to 0.999 yields the same performance as 0.2 (well above 95% acc). This is not possible, it means that the training/training switch inside the Dropout layer is broken, it is always returning the `inputs`, even during the training phase. 

With tensorflow 1.10, the tutorial script gives  the "correct" result which is ~11% accuracy with 99.9% droprate.

**Other info / logs**
It seems that either `tensorflow.keras.backend.learning_phase` is at the root of the problem, or `model.fit` doesn't correctly sets the training flag.

**Comment by feihugis** [comment]
@randolf-scholz I test it on tf-nightly(1.13.0.dev20190124). The results are in below and look correct.

When drop_rate = 0.99
```
Epoch 1/5
60000/60000==============================] - 5s 76us/sample - loss: 2.2957 - acc: 0.1542
Epoch 2/5
60000/60000==============================] - 4s 74us/sample - loss: 2.1568 - acc: 0.1891
Epoch 3/5
60000/60000==============================] - 4s 73us/sample - loss: 2.1143 - acc: 0.2052
Epoch 4/5
60000/60000==============================] - 4s 73us/sample - loss: 2.0843 - acc: 0.2174
Epoch 5/5
60000/60000==============================] - 4s 73us/sample - loss: 2.0794 - acc: 0.2218
10000/10000==============================] - 0s 24us/sample - loss: 1.1996 - acc: 0.8066
```
When drop_rate = 0.2
```
Epoch 1/5
60000/60000==============================] - 5s 84us/sample - loss: 0.2183 - acc: 0.9349
Epoch 2/5
60000/60000==============================] - 5s 82us/sample - loss: 0.0980 - acc: 0.9704
Epoch 3/5
60000/60000==============================] - 5s 79us/sample - loss: 0.0669 - acc: 0.9789
Epoch 4/5
60000/60000==============================] - 5s 80us/sample - loss: 0.0527 - acc: 0.9829
Epoch 5/5
60000/60000==============================] - 5s 79us/sample - loss: 0.0433 - acc: 0.9856
10000/10000==============================] - 0s 27us/sample - loss: 0.0697 - acc: 0.9801
```

**Comment by martinwicke** [comment]
@aselle I don't know when this was fixed, but this may be important enough to cherry-pick into 1.13 if it wasn't fixed before the branch cut.

**Comment by feihugis** [comment]
Also test in on ` tensorflow==1.13.0rc0`. The results look good.

**Comment by martinwicke** [comment]
Thanks @feihugis!

**Comment by fchollet** [comment]
@martinwicke tested dropout with 1.12 (from PyPI), works fine. We have unit tests for this, always had.

**Comment by martinwicke** [comment]
Hmm... strange thing to be broken by a conda rebuild. But then... who knows.

**Comment by randolf-scholz** [comment]
@fchollet @martinwicke 

The problems seems to appear in `tf.keras`, but not in the standalone `keras` package. The issue also seems affect the  `keras.backend.in_train_phase` phase switch. I tested it with the following files:

- [https://pastebin.com/SvTwrSC8](https://pastebin.com/SvTwrSC8)

- [https://pastebin.com/VtBwL5mz](https://pastebin.com/VtBwL5mz)

**Comment by ppwwyyxx** [comment]
@martinwicke tested dropout with 1.12 (from PyPI) and colab, with the above code. Does not works fine.

**Comment by rajatmonga** [comment]
Lets leave the issue open while we figure out more details and provide updates here.

**Comment by fchollet** [comment]
@randolf-scholz: thank you for taking the time to report this.

To be fair I have not done a full investigation and I have not looked at what was wrong with the script posted, or at which commit its behavior might have started changing.

Here is the test script I have used. I have only tested it with 1.13 and 1.12 on CPU. I have tested it in eager and graph mode. The script rigorously test that the Dropout layer injects ones in the forward pass activations of the Dense layer below it.

```python
import numpy as np
import tensorflow as tf

print('tf.version:', tf.__version__)  # 1.12.0

# tf.enable_eager_execution()  # try with eager or not

inputs = keras.Input(shape=(100,))
x = keras.layers.Dense(100,
                       kernel_initializer='zeros',
                       bias_initializer='ones',
                       trainable=False)(inputs)
x = keras.layers.Dropout(0.5)(x)
model = keras.Model(inputs, x)
model.compile(tf.train.RMSPropOptimizer(0.1), 'mse')

print(model.train_on_batch(np.ones((1, 100)), np.ones((1, 100))))  # Positive loss
print(model.fit(np.ones((1, 100)), np.ones((1, 100))))  # Positive loss

print(model.test_on_batch(np.ones((1, 100)), np.ones((1, 100))))  # 0 loss
print(model.evaluate(np.ones((1, 100)), np.ones((1, 100))))  # 0 loss
```

In addition to this, we have correctness tests for dropout in our test suite (multiple in fact, because we typically use a dropout layer as a way to test whether the learning phase was correctly set, which is something we have to test for a range of model types and situations).

To be the best I can tell tf.keras.layers.Dropout() is working normally in TF 1.12 in training and inference modes with `fit`, `train_on_batch`, `test_on_batch`, `evaluate`.

**Comment by fchollet** [comment]
I have now done a full investigation of the issue with the script posted. I confirm that incorrect behavior is present in 1.11 and 1.12.

This behavior is entirely unrelated to either the `Dropout` layer, or to the `in_train_phase` backend utility. The `Dropout` layer works completely fine.

The bug is an issue that occurs when using a `Sequential` model in "deferred mode". Deferred mode is a recently-introduce way to use `Sequential` without passing an `input_shape` argument as first layer. It's looking like the learning phase value was incorrectly set in this case.

I think this would not have affected many users since "deferred mode" for `Sequential` is new, and not widely used, because hardly any Keras example use it (none on keras.io). It is unfortunate that some of our tutorials on tensorflow.org have started using it.

I'm contacting deverel to make sure that the code examples are updated to add `input_shape` arguments in our Sequential models (which is better practice anyway since it allows for static layer compatibility checks).

You can fix the issue in the script by simply passing `input_shape=(28, 28)` to the first flatten layer.

This bug has been fixed some time ago on the TF side.

**Comment by epsi95** [comment]
I am creating the model below:
```
model = models.Sequential()
model.add(layers.Conv2D(32, (3,3), activation='relu', input_shape=(150,150,3)))
model.add(layers.MaxPool2D((2,2)))
model.add(layers.Conv2D(64, (3,3), activation='relu'))
model.add(layers.MaxPool2D((2,2)))
model.add(layers.Conv2D(128, (3,3), activation='relu'))
model.add(layers.MaxPool2D((2,2)))

model.add(layers.Flatten())
model.add(layers.Dropout(0.5))

model.add(layers.Dense(512, activation='relu'))
model.add(layers.Dense(1, activation='sigmoid'))

model.compile(loss='binary_crossentropy',
             optimizer=optimizers.RMSprop(lr=1e-4),
             metrics=['acc'])
```
The model summary says
```
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
=================================================================
conv2d_36 (Conv2D)           (None, 148, 148, 32)      896       
_________________________________________________________________
max_pooling2d_36 (MaxPooling (None, 74, 74, 32)        0         
_________________________________________________________________
conv2d_37 (Conv2D)           (None, 72, 72, 64)        18496     
_________________________________________________________________
max_pooling2d_37 (MaxPooling (None, 36, 36, 64)        0         
_________________________________________________________________
conv2d_38 (Conv2D)           (None, 34, 34, 128)       73856     
_________________________________________________________________
max_pooling2d_38 (MaxPooling (None, 17, 17, 128)       0         
_________________________________________________________________
flatten_12 (Flatten)         (None, 36992)             0         
_________________________________________________________________
dropout_11 (Dropout)         (None, 36992)             0         
_________________________________________________________________
dense_20 (Dense)             (None, 512)               18940416  
_________________________________________________________________
dense_21 (Dense)             (None, 1)                 513       
=================================================================
Total params: 19,034,177
Trainable params: 19,034,177
Non-trainable params: 0
```
Seems like dropout_11 (Dropout) layers don't have any effect. What am I doing wrong? 
-------------------------
keras version: '2.4.3'
tensorflow version: '2.3.1'
